flang/lib/Lower/NumericRuntime.cpp

```diff
@@ -0,0 +1,102 @@
+//===-- NumericRuntime.cpp -- runtime for numeric intrinsics -------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "../../runtime/numeric.h"
+#include "RTBuilder.h"
+#include "flang/Lower/Bridge.h"
+#include "flang/Lower/CharacterExpr.h"
+#include "flang/Lower/FIRBuilder.h"
+#include "flang/Lower/Support/BoxValue.h"
+#include "flang/Lower/Todo.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "flang/Lower/NumericRuntime.h"
+
+using namespace Fortran::runtime;
+
+/// Helper function to recover the KIND from the REAL FIR type.
+static int discoverRealKind(mlir::Type fltTy) {
+  if (fltTy.isF16())
+    return 2;
+  if (fltTy.isBF16())
+    return 3;
+  if (fltTy.isF32())
+    return 4;
+  if (fltTy.isF64())
+    return 8;
+  if (fltTy.isF80())
+    return 10;
+  if (fltTy.isF128())
+    return 16;
+  llvm_unreachable("unexpected type");
+}
+
+/// Generate call to RRSpacing intrinsic runtime routine. 
+mlir::Value
+Fortran::lower::genRRSpacing(Fortran::lower::FirOpBuilder &builder, 
+                           mlir::Location loc, mlir::Value x) { 
+  mlir::FuncOp func;
+  switch (discoverRealKind(x.getType())) {
+  case 4:
+    func = Fortran::lower::getRuntimeFunc<mkRTKey(RRSpacing4)>(loc, builder);
+    break;
+  case 8:
+    func = Fortran::lower::getRuntimeFunc<mkRTKey(RRSpacing8)>(loc, builder);
+    break;
+#if LONG_DOUBLE == 80
+  case 10:
+    func = Fortran::lower::getRuntimeFunc<mkRTKey(RRSpacing10)>(loc, builder);
+    break;
+#elif LONG_DOUBLE == 128
+  case 16:
+    func = Fortran::lower::getRuntimeFunc<mkRTKey(RRSpacing16)>(loc, builder);
+    break;
+#endif
+  default:
+    llvm_unreachable("unsupported real kind");
```

Comment:
Should be a `TODO(loc, "unsupported real kind in RRSpacing lowering")` since it is reachable for KIND 2, 3 and 10 or 16 (`llvm_unreachable` are removed in release mode)